tlog_filter.py

```diff
@@ -0,0 +1,103 @@
+"""
+
+import os
+import struct
+
+# Bug? I'm seeing mavlink.WIRE_PROTOCOL_VERSION == "1.0" for some QGC-generated tlog files
+# Force WIRE_PROTOCOL_VERSION to be 2.0
+os.environ['MAVLINK20'] = '1'
+
+import argparse
+
+from segment_reader import add_segment_args, choose_reader_list
+
+
+def filter_tlog(readers, output_filename, msg_types, sysid, compid, max_msgs, verbose):
+    """
+    Filter messages and write them to a new tlog file.
+    """
+    print(f'Writing {output_filename}')
+
+    with open(output_filename, 'wb') as outfile:
+        msg_count = 0
+        kept_count = 0
+
+        for reader in readers:
+            for msg in reader:
+                msg_count += 1
+
+                # Stop if we've hit the message limit
+                if max_msgs > 0 and kept_count >= max_msgs:
+                    print(f'Wrote {kept_count} messages, stopping')
+                    break
+
+                # Filter by sysid and compid
+                if sysid > 0 and sysid != msg.get_srcSystem():
+                    continue
+                if compid > 0 and compid != msg.get_srcComponent():
+                    continue
+
+                # Filter by message type
+                if msg_types is not None and msg.get_type() not in msg_types:
+                    continue
+
+                # We need to reconstruct the tlog header (8 bytes, Big Endian unsigned long long, microseconds)
+                header = struct.pack('>Q', int(msg._timestamp * 1_000_000))
+
+                # Write the header
+                outfile.write(header)
+
+                # Write the raw MAVLink packet data
+                outfile.write(msg.get_msgbuf())
+
+                kept_count += 1
+
+                if verbose and msg_count % 20000 == 0:
+                    print(f'... {msg_count} messages read, {kept_count} messages kept')
+
+            if max_msgs > 0 and kept_count >= max_msgs:
+                break
+
+        print(f'{msg_count} messages read, {kept_count} messages kept')
+
+
+def main():
+    parser = argparse.ArgumentParser(formatter_class=argparse.RawDescriptionHelpFormatter, description=__doc__)
+    add_segment_args(parser)
+    parser.add_argument('-v', '--verbose', action='store_true',
+                        help='print a lot more information')
+    parser.add_argument('--types', default=None,
+                        help='comma separated list of message types, e.g., "ATTITUDE,RC_CHANNELS"')
+    parser.add_argument('--sysid', type=int, default=0,
+                        help='select source system id, default is all systems')
+    parser.add_argument('--compid', type=int, default=0,
+                        help='select source component id, default is all components')
+    parser.add_argument('--max-msgs', type=int, default=500000,
+                        help='stop after keeping this number of messages, default is 500K')
+    parser.add_argument('--output', default='filtered.tlog',
+                        help='output file name')
```

Comment:
No need for the --output argument.

Segment naming works like this:
filename = util.get_outfile_name(reader.name, suffix='_filtered')